idea/src/org/jetbrains/kotlin/idea/intentions/JoinDeclarationAndAssignmentIntention.kt

```diff
@@ -76,25 +77,33 @@ class JoinDeclarationAndAssignmentIntention : SelfTargetingRangeIntention<KtProp
     }
 
     override fun applyTo(element: KtProperty, editor: Editor?) {
-        val typeReference = element.typeReference ?: return
-
+        if (element.typeReference == null) return
         val assignment = findAssignment(element) ?: return
         val initializer = assignment.right ?: return
-        val newInitializer = element.setInitializer(initializer)!!
 
+        element.initializer = initializer
         if (element.hasModifier(KtTokens.LATEINIT_KEYWORD)) element.removeModifier(KtTokens.LATEINIT_KEYWORD)
 
-        val initializerBlock = assignment.parent.parent as? KtAnonymousInitializer
-        assignment.delete()
-        if (initializerBlock != null && (initializerBlock.body as? KtBlockExpression)?.isEmpty() == true) {
-            initializerBlock.delete()
+        val grandParent = assignment.parent.parent
+        val initializerBlock = grandParent as? KtAnonymousInitializer
+        val secondaryConstructor = grandParent as? KtSecondaryConstructor
+        val newProperty = if (initializerBlock != null || secondaryConstructor != null) {
+            assignment.delete()
```

Comment:
In this branch, *two* different cases are possible, and it seems to me you consider only one of them:

- *local variable* declared *and* assigned in anonymous initializer / secondary constructor, in this case result should be placed at the assignment
- *property* of the class which is assigned in anonymous initializer / secondary constructor, in this case result should be placed at the declaration